src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -256,90 +267,108 @@ private static boolean shouldBeDeclaredAsFinal(ClassDesc desc) {
+                .ifPresent(ClassDesc::registerNestedSubclass);
         }
     }
 
     /**
-     * Checks if there is a class with same name.
+     * Register to the super class of anonymous inner class that the given class is instantiated
+     * by an anonymous inner class.
+     *
+     * @param literalNewAst ast node of {@link TokenTypes#LITERAL_NEW} representing anonymous inner
+     *                      class
+     * @param outerTypeDeclName Fully qualified name of the outer type declaration of anonymous
+     *                          inner class
+     */
+    private void registerAnonymousInnerClassToSuperClass(DetailAST literalNewAst,
+                                                         String outerTypeDeclName) {
+        final String superClassName = CheckUtil.getShortNameOfAnonInnerClass(literalNewAst);
+
+        final Function<ClassDesc, Integer> anonClassCountProvider = classDesc -> {
+            return getAnonSuperTypeMatchingCount(outerTypeDeclName, classDesc.getQualifiedName());
+        };
+        getNearestClassWithSameName(superClassName, anonClassCountProvider)
+            .or(() -> Optional.ofNullable(innerClasses.get(superClassName)))
+            .ifPresent(ClassDesc::registerSuperClassOfAnonymousInnerClass);
+    }
+
+    /**
+     * Get the nearest class with same name.
+     *
+     * <p>The parameter {@code countProvider} exists because if the class being searched is the
+     * super class of anonymous inner class, the rules of evaluation are a bit different,
+     * consider the following example-
+     * <pre>
+     * {@code
+     * public class Main {
+     *     static class One {
+     *         static class Two {
+     *         }
+     *     }
+     *
+     *     class Three {
+     *         One.Two object = new One.Two() { // Object of Main.Three.One.Two
+     *                                          // and not of Main.One.Two
+     *         };
+     *
+     *         static class One {
+     *             static class Two {
+     *             }
+     *         }
+     *     }
+     * }
+     * }
+     * </pre>
+     * If the {@link Function} {@code countProvider} hadn't used
+     * {@link FinalClassCheck#getAnonSuperTypeMatchingCount} to
+     * calculate the matching count then the logic would have falsely evaluated
+     * {@code Main.One.Two} to be the super class of the anonymous inner class.
      *
      * @param className name of the class
-     * @param superClassName name of the super class
-     * @return true if there is another class with same name.
+     * @param countProvider the function to apply to calculate the name matching count
+     * @return {@link Optional} of {@link ClassDesc} object of the nearest class with the same name.
      * @noinspection CallToStringConcatCanBeReplacedByOperator
      */
-    private ClassDesc getNearestClassWithSameName(String className, String superClassName) {
+    private Optional<ClassDesc> getNearestClassWithSameName(String className,
```

Comment:
@Vyom-Yadav I think we can still use fancy syntax as you proposed before, with no negatives in intuitive manner:

```suggestion
    private Optional<ClassDesc> getNearestClassWithSameName(String className,
                                                            Function<ClassDesc, Integer> countProvider) {
        final String dotAndClassName = PACKAGE_SEPARATOR.concat(className);
        final Comparator<ClassDesc> longestMatch = Comparator.comparingInt(countProvider::apply);

        return innerClasses.entrySet().stream()
                           .filter(entry -> entry.getKey().endsWith(dotAndClassName))
                           .map(Map.Entry::getValue)
                           .min(longestMatch.reversed().thenComparingInt(ClassDesc::getDepth));
    }
```

Please try this.

Reply:
Thanks for the suggestion, implemented. This Github feature is also nice.

Reply:
Resolved.